src/macros-by-example.adoc

```diff
@@ -462,25 +465,42 @@ types are verified:
 TODO: demonstrate splicing in TDL macro invocations
 
 
-=== Mapping Templates Over Streams
+=== Mapping Templates Over Streams: `*for*`
 
-Another way to produce a stream is via a mapping form. The `*for_each*` special form evaluates a
-template once for each value provided by a macro parameter.  Each time, the parameter name is
-shadowed and rebound to the next value on the stream.
+Another way to produce a stream is via a mapping form.  The `*for*` special form evaluates a
+template once for each value provided by a stream or streams.  Each time, a local variable is
+created and bound to the next value on the stream.
 
 [{mrk}]
 ----
 (*public* prices
-  [(*symbol* currency), (*number\...* amount)]
-  (*for_each* amount
-    (price amount currency)))
+  [(*symbol* currency), (*number\...* amounts)]
+  (*for* [(amt amounts)]
```

Comment:
Would you consider `[amt::amounts]`, `[(amt in amounts)]`, or some other syntax so that it doesn't look quite so much like `[(symbol currency)]`?

I don't think there's a lot of value in being terse in the template expressions for Ion text encoding. Adding a few extra words is a small price to pay for readability (and by extension, ease of learning/use).

Reply:
TBH as soon as this example came together I thought that its the _signature_ that's the problem here.

This is following longstanding Lisp convention, in fact this syntax is [lifted from Racket](https://docs.racket-lang.org/reference/for.html). I'm inclined to continue that tradition, because it's proven itself worthy in terms of extensibility (meaning there's lots of syntactic "room" to add new capabilities to existing forms) and consistency (meaning similarly-behaving forms will have similar syntax).  Relevant here is the convention that variable declarations (ie binding forms) put the variable name first, followed by the expression to be evaluated.

The current signature syntax goes against this principle, following C-like conventions. If the order of macro parameter-specs were reversed (eg strawman `[(currency symbol)]` then these would be consistent in following in idiom that's more likely to be usable for the other forms we need to add.

Reply:
What I was trying to say is that the `(inner_binding expression)` syntax looks an awful lot like the `(type param_name)` syntax (regardless of order). The fact that both are a list of 2-element s-expressions means that they look similar. Adding a keyword—or _anything_—to the text format of the `for` binding (e.g. `(inner_binding of outer_var)`) provides a local contextual clue. Sure, it's not _necessary_, but I'm thinking from an ease-of-use perspective.

I'm not trying to advocate for any particular syntax here—I just think it's worth considering the tradeoffs between purity/minimalism and approachability.

Reply:
I get that, and will close by saying that the similar shape is an intentional syntactic pattern.  It's safe to say that different folks have different feelings about whether the minimalism is a benefit or a drawback.

Since all of this syntax is provisional and nothing's been implemented, I ask you bear with your (reasonable!) syntactic reservations as more of the language comes up for review. If things still don't gel in your mind, we can have the discussion with the full grammar at hand.